File: JavaScript/incrementManager2.js
/**
 * incrementManager2 uses Decorators to simplify the inheritance
 * hierarchy.
 */



// ==================================================================
// START: IncrementManager Interface







function IIncrementManager() {
	'use strict';
}
IIncrementManager.prototype = new Object();
IIncrementManager.prototype.constructor = IIncrementManager;

IIncrementManager.prototype.step = 0;
IIncrementManager.prototype.isCumulative = false;
IIncrementManager.prototype.isInfinite = false;
IIncrementManager.prototype.doesReset = false;
IIncrementManager.prototype.doesOscillate = false;

IIncrementManager.prototype.updateStep = function () {
	'use strict';
};

IIncrementManager.prototype.getStep = function () {
	'use strict';
	return this.step;
};
IIncrementManager.prototype.getSteps = function () {
	'use strict';
};
IIncrementManager.prototype.getIncrement = function () {
	'use strict';
	return 0;
};
IIncrementManager.prototype.getMin = function () {
	'use strict';
	return this.step;
};
IIncrementManager.prototype.getMax = function () {
	'use strict';
	return this.step;
};
IIncrementManager.prototype.getLastStep = function () {
	'use strict';
	return this.step;
};







//  END:  IncrementManager interface
// ==================================================================
// START: IncrementDecorator interface







function IIncrementDecorator(IncMan) {
	'use strict';
}
//IIncrementDecorator.prototype = Object.create(IIncrementManager.prototype);
IIncrementDecorator.prototype = new IIncrementManager();
IIncrementDecorator.prototype.constructor = IIncrementDecorator;
IIncrementDecorator.prototype.incMgr = null;
IIncrementDecorator.prototype.lastStep = 0;
IIncrementDecorator.prototype.getLastStep = function () {
	'use strict';
	return this.lastStep;
};

IIncrementDecorator.prototype.getIncrement = function () {
	'use strict';
	return this.lastStep - this.step;
};
IIncrementDecorator.prototype.getSteps = function () {
	'use strict';
	return this.incMgr.getSteps();
};







//  END:  IncrementDecorator interface
// ==================================================================
// START: IncrementManager







function IncrementManager(step) {
	'use strict';
	var msg = '';
	if (typeof step !== 'number') {
		msg = 'IncrementManager constructor expects only parameter "step" to be a number';
		console.error(msg);
		throw {"message": msg};
	}
	this.step = step;
}
//IncrementManager.prototype = Object.create(IIncrementManager.prototype);
IncrementManager.prototype = new IIncrementManager();
IncrementManager.prototype.constructor = IncrementManager;
IncrementManager.prototype.steps = 0;

IncrementManager.prototype.updateStep = function () {
	'use strict';
	this.steps += 1;
};
IncrementManager.prototype.getSteps = function () {
	'use strict';
	return this.steps;
};







//  END:  IncrementManager
// ==================================================================
// START: IncrementDecoratorDecay







function IncrementDecoratorDecay(IncMan, decay) {
	'use strict';
	var msg = 'IncrementDecoratorDecay constructor expects ';
	if (!IncMan instanceof IIncrementManager) {
		msg += 'first parameter to be an instance of IIncrementManager.';
		console.error(msg);
		throw {"message": msg};
	}

	if (typeof decay !== 'number') {
		msg += 'second parameter "decay" to be a number';
		console.error(msg);
		throw {"message": msg};
	}

	this.incMgr = IncMan;
	this.decayFactor = decay;
}
//IncrementDecoratorDecay.prototype = Object.create(IIncrementDecorator.prototype);
IncrementDecoratorDecay.prototype = new IIncrementDecorator();
IncrementDecoratorDecay.prototype.constructor = IncrementDecoratorDecay;

IncrementDecoratorDecay.prototype.isCumulative = false;
IncrementDecoratorDecay.prototype.isInfinite = false;
IncrementDecoratorDecay.prototype.doesOscillate = false;
IncrementDecoratorDecay.prototype.doesReset = false;
IncrementDecoratorDecay.prototype.decayFactor = 0;
IncrementDecoratorDecay.prototype.decayValue = 1;

IncrementDecoratorDecay.prototype.getMin = function () {
	'use strict';
	if (this.decayFactor < 1 && this.decayFactor > -1) {
		return 0;
	} else {
		return this.incMgr.getStep();
	}
};
IncrementDecoratorDecay.prototype.getMax = function () {
	'use strict';
	if (this.decayFactor < 1 && this.decayFactor > -1) {
		return this.incMgr.getStep();
	} else {
		return false;
	}
};
IncrementDecoratorDecay.prototype.updateStep = function () {
	'use strict';
	this.decayValue *= this.decayFactor;
	this.lastStep = this.step;
	this.incMgr.updateStep();
	this.step = this.decayValue * this.incMgr.getStep();
};







//  END:  IncrementDecoratorDecay
// ==================================================================
// START: IncrementDecoratorCumulative







function IncrementDecoratorCumulative(IncMan) {
	'use strict';
	var msg = '';
	if (!IncMan instanceof IIncrementManager) {
		msg = 'IncrementDecoratorCumulative constructor expects first parameter to be an instance of IIncrementManager.';
		console.error(msg);
		throw {"message": msg};
	}
	this.incMgr = IncMan;
}
//IncrementDecoratorCumulative.prototype = Object.create(IIncrementDecorator.prototype);
IncrementDecoratorCumulative.prototype = new IIncrementDecorator();
IncrementDecoratorCumulative.prototype.constructor = IncrementDecoratorCumulative;
IncrementDecoratorCumulative.prototype.isCumulative = true;
IncrementDecoratorCumulative.prototype.isInfinite = true;
IncrementDecoratorCumulative.prototype.doesOscillate = false;
IncrementDecoratorCumulative.prototype.doesReset = false;
IncrementDecoratorCumulative.prototype.updateStep = function () {
	'use strict';
	this.incMgr.updateStep();
	this.step += this.incMgr.getStep();
};
IncrementDecoratorCumulative.prototype.getMax = function () {
	'use strict';
	return false;
};







//  END:  IncrementDecoratorCumulative
// ==================================================================
// START: IncrementDecoratorReset







function IncrementDecoratorReset(IncMan, min, max, preset) {
	'use strict';

	var tmp = 0,
		tmp2 = 0,
		msg = this.constructor.name + ' constructor expects ';

	if (!IncMan instanceof IIncrementManager) {
		msg += 'first parameter to be an instance of IIncrementManager.';
		console.error(msg);
		throw {"message": msg};
	}
	this.incMgr = IncMan;
	if (!min instanceof IIncrementManager) {
		msg += 'second parameter "min" to be an instance of IIncrementManager.';
		console.error(msg);
		throw {"message": msg};
	}
	if (!max instanceof IIncrementManager) {
		msg += 'third parameter "max" to be an instance of IIncrementManager.';
		console.error(msg);
		throw {"message": msg};
	}
	if (min.getStep() > max.getStep()) {
		msg += 'second parameter min.getStep() to be less than third parameter max.getStep()';
		console.error(msg);
		throw {"message": msg};
	}

	tmp = max.getStep() - min.getStep();
	tmp2 = IncMan.getStep();
	if (tmp2 < 0) {
		tmp2 = -tmp2;
	}
	if (tmp2 > tmp) {
		msg += 'first parameter "IncMan" (' + IncMan.getStep() + ') to be a number that can be made to fit within ' + min.getStep() + ' and ' + max.getStep();
		console.error(msg);
		throw {"message": msg};
	}

	this.min = min;
	this.max = max;

	if (preset !== undefined) {
		if (typeof preset !== 'number' || preset < 0 || preset > 1) {
			msg += 'fourth parameter "preset" to be a number between 0 and 1 (inclusive).';
			console.error(msg);
			throw {"message": msg};
		} else {
			this.step = this.min.getStep() + ((this.max.getStep() - this.min.getStep()) * preset);
		}
	} else {
		this.step = min.getStep();
	}
}
//IncrementDecoratorCumulative.prototype = Object.create(IIncrementDecorator.prototype);
IncrementDecoratorReset.prototype = new IIncrementDecorator();
IncrementDecoratorReset.prototype.constructor = IncrementDecoratorReset;
IncrementDecoratorReset.prototype.min = null;
IncrementDecoratorReset.prototype.max = null;
IncrementDecoratorReset.prototype.isCumulative = true;
IncrementDecoratorReset.prototype.doesReset = true;
IncrementDecoratorReset.prototype.cumulative = 0;

IncrementDecoratorReset.prototype.updateStep = function () {
	'use strict';
	this.lastStep = this.step;

	this.incMgr.updateStep();
	this.min.updateStep();
	this.max.updateStep();
	this.step += this.incMgr.getStep();

	if (this.step === this.max.getStep()) {
		this.step = this.min.getStep();
	} else if (this.step === this.min.getStep()) {
		this.step = this.max.getStep();
	} else if (this.step > this.max.getStep()) {
		this.step = this.min.getStep() + this.step - this.max.getStep();
	} else if (this.step < this.min.getStep()) {
		this.step = this.max.getStep() - (this.min.getStep() - this.step);
	}
};
IncrementDecoratorReset.prototype.getMin = function () {
	'use strict';
	this.max.getStep();
};
IncrementDecoratorReset.prototype.getMax = function () {
	'use strict';
	this.min.getStep();
};
IncrementDecoratorReset.prototype.getLastStep = function () {
	'use strict';
	this.lastStep;
};







//  END:  IncrementDecoratorReset
// ==================================================================
// START: IncrementDecoratorOscillate







function IncrementDecoratorOscillate(IncMan, min, max, preset) {
	'use strict';
	IncrementDecoratorReset.call(this, IncMan, min, max, preset);
//	this.constructorFuncShared(IncMan, min, max, preset);
}
//IncrementDecoratorCumulative.prototype = Object.create(IncrementDecoratorReset.prototype);
IncrementDecoratorOscillate.prototype = new IncrementDecoratorReset(new IncrementManager(0), new IncrementManager(0), new IncrementManager(0));
IncrementDecoratorOscillate.prototype.constructor = IncrementDecoratorOscillate;
IncrementDecoratorOscillate.prototype.isCumulative = true;
IncrementDecoratorOscillate.prototype.isInfinite = false;
IncrementDecoratorOscillate.prototype.doesOscillate = true;
IncrementDecoratorOscillate.prototype.doesReset = false;
IncrementDecoratorOscillate.prototype.cumulative = 0;
IncrementDecoratorOscillate.prototype.lastCumulative = 0;
IncrementDecoratorOscillate.prototype.changeSign = false;


IncrementDecoratorOscillate.prototype.updateStep = function () {
	'use strict';
	var step;
	this.lastCumulative = this.cumulative;
	this.incMgr.updateStep();
	this.min.updateStep();
	this.max.updateStep();
	this.lastStep = this.step;

	this.step = this.incMgr.getStep();
	if (this.changeSign === true) {
		this.step = -this.step;
	}

	this.cumulative += this.step;


	if (this.cumulative > this.max.getStep()) {
		this.step = this.max.getStep() - this.cumulative;
		this.cumulative = this.max.getStep() - this.step;
		this.changeSign = true;
	} else if (this.cumulative < this.min.getStep()) {
		this.step = this.cumulative - this.min.getStep();
		this.cumulative = this.min.getStep() + this.step;
		this.changeSign = false;
	}
};


IncrementDecoratorOscillate.prototype.getStep = function () {
	'use strict';
	return this.step;
};





//  END:  IncrementDecoratorOscillate
// ==================================================================

File: ArduinoDriver/Increment-Manager-2.cpp
#include "Increment-Manager.h"


// ==================================================================
// START: IncrementManager Interface







bool IIncrementManager::isCumulative() {
	return _isCumulative;
};
bool IIncrementManager::isInfinite() {
	return _isInfinite;
}
bool IIncrementManager::doesOscillate() {
	return _doesOscillate;
};

void IIncrementManager::_correctMinMax( double &min, double &max) {
	double tmp;
	if( min > max )
	{
		tmp = min;
		min = max;
		max = tmp;
	}
}







//  END:  IncrementManager Interface
// ========================================================
// START: IncrementDecorator Abstract








unsigned int AIncrementDecorator::getStepCount() {
	return incMgr.getStepCount();
}




//  END:  IncrementDecorator Abstract
// ========================================================
// START: IncrementManager





IncrementManager::IncrementManager(double step) {
	_step = step;
};
/// makes increment change (if it can)
/// must be called once for every loop
void IncrementManager::updateStep() {
	_stepCount += 1;
};

/// returns the current step value
double IncrementManager::getStep() {
	return _step;
};

/// returns the number of steps the increment manager has done
unsigned int IncrementManager::getStepCount() {
	return _stepCount;
};

/// returns the current increment value
double IncrementManager::getIncrement() {
	return _step;
};

/// returns the previous step
double IncrementManager::getLastStep() {
	return _step;
};

/// returns the minimum value the step/increment/cumulative can be
double IncrementManager::getMin() {
	return _step;
};

/// returns the maximum value the step/increment/cumulative can be
double IncrementManager::getMax() {
	return _step;
};

bool IncrementManager::withinMinMax(double min, double max) {
	_correctMinMax(min,max);

	if( _step >= min && _step <= max )
	{
		return true;
	}
	else
	{
		return false;
	}
};






//  END:  IncrementManager
// ========================================================
// START: IncrementDecoratorAccumulate





IncrementDecoratorAccumulate::IncrementDecoratorAccumulate( IIncrementManager incMan ) {
	_incMgr = incMan
}
IncrementDecoratorAccumulate::IncrementDecoratorAccumulate( IIncrementManager incMan , double preset ) {
	_incMgr = incMan
	_cumulative = preset;
}
void IncrementDecoratorAccumulate::updateStep() {
	_incMgr.updateStep();
	_lastStep = _cumulative;;
	_cumulative += _incMgr.getStep();
}
double IncrementDecoratorAccumulate::getStep() {
	return _cumulative;
}
double IncrementDecoratorAccumulate::getIncrement() {
	return _incMgr.getStep();
}
double IncrementDecoratorAccumulate::getLastStep() {
	return _lastStep;
}
bool IncrementDecoratorAccumulate::withinMinMax() {
	return false;
}
double IncrementDecoratorAccumulate::getMin() {
	// this requires much more thought (and is possibly impossible to work out)
	//if(incMgr.getStep() > 0)
}
double IncrementDecoratorAccumulate::getMax() {
	// this requires much more thought (and is possibly impossible to work out)
	//if(incMgr.getStep() > 0)
}





//  END:  IncrementDecoratorAccumulate
// --------------------------------------------------------
// START: IncrementDecoratorAccumulatorReset





IncrementDecoratorAccumulatorReset::IncrementDecoratorAccumulatorReset(IIncrementManager incMan, IIncrementManager min , IIncrementManager max) {
	_incMgr = incMan;
	_setCorrectMinMax(min,max);
	_cumulative = _min.getStep();
}
IncrementDecoratorAccumulatorReset::IncrementDecoratorAccumulatorReset(IIncrementManager IncMan, IIncrementManager min , IIncrementManager max, double preset) {
	_incMgr = IncMan;
	_setCorrectMinMax(min,max);
	_cumulative = preset;
}
void IncrementDecoratorAccumulatorReset::updateStep() {
	_incMgr.updateStep();
	_min.updateStep();
	_max.updateStep();

	_lastStep = _step;

	double min = _min.getStep();
	double max = _max.getStep();

	_cumulative += _incMgr.getStep();

	if( _cumulative > max )
	{
		_cumulative = min + (_cumulative - max);
	}
	else
	{
		if( _cumulative < min )
		{
			_cumulative = max - (min - _cumulative);
		}
	}
}

double IncrementDecoratorAccumulatorReset::getStep() {
	return _cumulative;
}

bool IncrementDecoratorAccumulatorReset::withinMinMax(double min, double max) {
	_correctMinMax(min, max);
	if(_min.getMin() >= min && _max.getMax() <= max)
	{
		return true;
	}
	else
	{
		return false;
	}
}

/// returns the minimum value the step/increment/cumulative can be
double IncrementDecoratorAccumulatorReset::getMin() {
	return _min.getStep();
};

/// returns the maximum value the step/increment/cumulative can be
double IncrementDecoratorAccumulatorReset::getMax() {
	return _max.getStep();
};


void IncrementDecoratorAccumulatorReset::_setCorrectMinMax(double min, double max) {
	if( min.getStep() > max.getStep() )
	{
		_min = max;
		_max = min;
	}
	else
	{
		_min = min;
		_max = max;
	}
};





//  END:  IncrementDecoratorAccumulatorReset
// ========================================================
// START: IncrementDecoratorAccumulatorOscillate



void IncrementDecoratorAccumulatorOscillate::updateStep() {
	_incMgr.updateStep();
	_min.updateStep();
	_max.updateStep();

	_lastStep = _cumulative;

	double min = _min.getStep();
	double max = _max.getStep();

	_cumulative += _incMgr.getStep();

	if( _cumulative > max )
	{
		_cumulative = min + (_cumulative - max);
	}
	else
	{
		if( this._cumulative < min )
		{
			this._cumulative = max - (min - this._cumulative);
		}
	}
}

double IncrementDecoratorAccumulatorOscillate::getStep() {
	return this._cumulative;
}
double IncrementDecoratorAccumulatorOscillate::getIncrement() {
	return _step - _lastStep;
}


//  END:  IncrementDecoratorAccumulatorOscillate
// ========================================================
// START: IncrementDecoratorDecay



IncrementDecoratorDecay::IncrementDecoratorDecay(IIncrementManager incMan, double decayFactor) {
	_incMgr = incMan;
	_decayFactor = decayFactor;
	_lastStep = _incMgr.getStep();
	_step = _incMgr.getStep();
}
double IncrementDecoratorDecay::updateStep() {
	_incMgr.updateStep();
	_lastStep = _step;
	_decayValue *= _decayFactor;
	_step *= _decayValue;
}

double IncrementDecoratorDecay::getStep() {
	return _step;
}

bool IncrementDecoratorDecay::withinMinMax(double min, double max) {
	_correctMinMax(min,max);
	if( _decayFactor <= 1 && _decayFactor >= 0 && _incMgr.getStep() <= max && _incMgr.getStep() >= min && min <= 0 )
	{
		return true;
	}
	return false;
}

/// returns the minimum value the step/increment/cumulative can be
double IncrementDecoratorDecay::getMin() {
	double tmp = incMgr.getStep();
	if( _decayFactor <= 1 && _decayFactor >= 0 )
	{
		if(tmp > 0)
		{
			return 0;
		} else {
			return tmp;
		}
	}
	else if(_decayFactor >= -1 && _decayFactor <= 0)
	{
		return -tmp;
	}
	else if(_decayFactor > 1)
	{
		return tmp;
	}
	else
	{
		// _decayFactor must be less than -1
		// throw // not sure what to do here
	}
}

/// returns the maximum value the step/increment/cumulative can be
double IncrementDecoratorDecay::getMax() {
	double tmp = incMgr.getStep();
	if( _decayFactor <= 1 && _decayFactor >= 0 )
	{
		if(tmp > 0)
		{
			return tmp;
		} else {
			return 0;
		}
	}
	else if(_decayFactor >= -1 && _decayFactor <= 0)
	{
		return tmp;
	}
	else if(_decayFactor > 1)
	{
		if( tmp < 0 )
		{
			return tmp;
		}
		else
		{
			// throw // not sure what to do here
		}
	}
	else
	{
		// _decayFactor must be less than -1
		// throw // not sure what to do here
	}
}


double IncrementDecoratorDecay::getIncrement() {
	return _step - _lastStep;
}


//  END:  IncrementDecoratorDecay
// ========================================================